aten/src/ATen/NamedTensor.cpp

```diff
@@ -1,14 +1,37 @@
 #ifdef NAMEDTENSOR_ENABLED
 #include <ATen/NamedTensor.h>
+#include <ATen/core/Tensor.h>
+#include <torch/csrc/utils/memory.h>
 
 namespace at {
 
 bool NamedTensorMeta::has_names() const {
   return !std::all_of(
-      names.begin(), names.end(), [](const Dimname& n) {
+      names_.begin(), names_.end(), [](const Dimname& n) {
         return n.type() == NameType::WILDCARD;
       });
 }
 
+void internal_set_names_inplace(Tensor tensor, optional<DimnameList> names) {
+  if (!names) {
+    tensor.unsafeGetTensorImpl()->set_named_tensor_meta(nullptr);
+    return;
+  }
+
+  auto ndim = tensor.dim();
+  TORCH_CHECK(ndim == names->size(),
+      "Number of names (", names->size(), ") and "
+      "number of dimensions in tensor (", ndim, ") ",
+      "do not match.");
+
+  auto* meta = tensor.get_named_tensor_meta();
+  if (meta == nullptr) {
+    tensor.unsafeGetTensorImpl()->set_named_tensor_meta(
+        torch::make_unique<NamedTensorMeta>(*names));
```

Comment:
another c++ question... is this going to copy the vector of names into a new vector? does the pass by value also copy (copying the list twice?). 

Would it be any better to pass the tensor by reference?

Reply:
`DimnameList` is defined to be `ArrayRef<Dimname>`. You can think of it like a `reference_wrapper<vector<Dimname>>` (so yes, it's already being passed by reference).

Reply:
ah OK. And the Tensor argument?

Reply:
It's probably slightly faster to pass `Tensor` as `Tensor&`. `Tensor` itself is an `intrusive_ptr<TensorImpl>` I think; passing it as `Tensor` means we bump the refcount and copy the pointer (which is fast, because it's just a word) and passing it as `Tensor&` means there is no refcount bump.

Reply:
I ask partly because it seems like we usually use `Tensor &`:
```
(base) [bvaughan@devgpu005.ash6 ~/repos/pytorch] gg -rnE '\(Tensor & [a-zA-Z]+.*\).*;' .  | grep \\.h   | wc -l
3941
(base) [bvaughan@devgpu005.ash6 ~/repos/pytorch] gg -rnE '\(Tensor [a-zA-Z]+.*\).*;' .  | grep \\.h   | wc -l
51
```

Reply:
Yup, I'll change it to `const Tensor&` (I think that is what we use for functions that modify tensors)